src/tools/perf/perftest_mad.c

```diff
@@ -0,0 +1,697 @@
+    if (!umad) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ucs_debug("MAD: receive: Allocated len:%zu", ref_len);
+retry:
+    if (ref_len > INT_MAX - umad_size()) {
+        free(umad);
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    user_mad = umad;
+    len      = ref_len;
+
+    ret = umad_recv(fd, umad, &len, timeout);
+    if (ret < 0) {
+        if (errno == ETIMEDOUT || errno == EAGAIN) {
+            free(umad);
+            return UCS_ERR_NO_PROGRESS;
+        }
+        if (errno != EINVAL && errno != ENOSPC) {
+            ucs_error("MAD: receive: Failed, umad len:%d, ret:%d, errno:%d",
+                      len, ret, errno);
+            free(umad);
+            return UCS_ERR_IO_ERROR;
+        }
+        /* EINVAL: no length info, ENOSPC: needed length is set */
+        if (errno == EINVAL) {
+            ref_len *= 2; /* 'len' has become invalid */
+        } else {
+            ref_len = len;
+        }
+        umad = realloc_or_free(umad, umad_size() + ref_len);
+        if (!umad) {
+            return UCS_ERR_NO_MEMORY;
+        }
+        goto retry;
+    }
+
+    if (perftest_mad_get_remote_port(umad, remote_port) != UCS_OK) {
+        free(umad);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    len -= IB_VENDOR_RANGE2_DATA_OFFS;
+    if (len <= 0) {
+        if (user_mad->status == ETIMEDOUT) {
+            ucs_info("MAD: receive: Remote unreachable");
+            free(umad);
+            return UCS_ERR_UNREACHABLE;
+        }
+        free(umad);
+        return UCS_ERR_OUT_OF_RANGE;
+    }
+
+    ret = umad_status(umad);
+    if (ret) {
+        ucs_info("MAD: receive: Status failure: %d", ret);
+        free(umad);
+        return UCS_ERR_REJECTED;
+    }
+    ret = UCS_OK;
+
+    data = (uint8_t*)umad_get_mad(umad) + IB_VENDOR_RANGE2_DATA_OFFS;
+    if (perftest_mad_is_ping(data, len)) {
+        ucs_debug("PONG");
+        goto retry;
+    }
+
+    if (len > *avail) {
+        ret = UCS_ERR_MESSAGE_TRUNCATED;
+        len = *avail;
+    }
+    memcpy(buffer, data, len);
+    *avail = len;
+    free(umad);
+    return ret;
+}
+
+static ucs_status_t
+perftest_mad_recv_from_remote(perftest_mad_rte_group_t *rte_group, void *buffer,
+                              size_t *avail, const ib_portid_t *target_port)
+{
+    ucs_status_t ret        = UCS_ERR_IO_ERROR;
+    ib_portid_t remote_port = {
+        .lid = 0
+    };
+    size_t size;
+
+    while (!remote_port.lid || remote_port.lid != target_port->lid) {
+        size            = *avail;
+        remote_port.lid = 0;
+
+        ret = perftest_mad_recv(rte_group, buffer, &size, &remote_port);
+        if (ret == UCS_ERR_NO_PROGRESS) {
+            ucs_info("PING");
+            ret = perftest_mad_ping(rte_group);
+            ucs_assert(ret == UCS_OK);
+            remote_port.lid = 0;
+            continue;
+        }
+        if (ret == UCS_ERR_UNREACHABLE) {
+            ucs_error("MAD: recv: Remote unreachable");
+            exit(EXIT_FAILURE);
```

Comment:
do we really need to exit here? Maybe better return an error and handle it later?

Reply:
It is perftest specific, if we get this error, we don't have a remote anymore, so it seems more reliable to bail, as done for socket rte.

Reply:
can we return some error code here, which we can check later and exit if needed?